### ISSUE TITLE
IDLE COMMIT connections on postgresql

### ISSUE THREAD

**Issue by Rastopapola** [pallets/werkzeug#1854]
# Bug
## Background
I found this rather by accident but it might be a bigger problem. I'm working on a Django3 project, using postgresql and the latest version of Werkzeug (1.0.1). It looks like the response handling via `runserver_plus` does not finish it's work properly, when a new request is already incoming. 
I assumed we had a very bad bug in our logic but found the source outside of our code.

Further details below:

## Steps to reproduce
Assume you have a running postgresql database with tables and everything is nicely set up.
1. Take a look on active db connections via 
    1. ```
        su - postgres
        psql
        SELECT queries FROM pg_stat_activity;
        ```
1. Run your django server using 
    1. ```
        python manage.py runserver_plus
        ```
1.  Navigate on a route, which has any form of db fetching and populating. It is enough to add the following to a test route
     1. ```
         MODEL.objects.all().first()
        ```
1. Refresh your browser like a maniac by holding F5 for about two seconds.
1. Check again your `pg_stat_acitvity`.
    1. There are a lot of idle connections, which queries are only set to `COMMIT` and they won't be killed.

## What normally happens
The opened connection will be closed after this part of code, which can be found in `werkzeug/debug/__init__.py`:
```python
l. 300 following
    def debug_application(self, environ, start_response):
        """Run the application and conserve the traceback frames."""
        app_iter = None
        try:
            app_iter = self.app(environ, start_response)
            for item in app_iter:
                yield item
            if hasattr(app_iter, "close"):
                app_iter.close()
    ...
```
If 'app_iter.close()' is reached, everything works as expected.

## What happened for us
But if we refresh extremely fast, the response handling seems to be interrupted somewhere in the middleware chain of Django and the new request instead is being handled, leaving the open connection as it is.

We really like `Werkzeug`, which is why we use it for two years now and it's the reason I want to inform you on this behaviour. Maybe you are aware of this, maybe it's really an unknown bug - I don't know.

Hope that helps.

**Comment by davidism** [comment]
Thank you for the detailed report. However, I don't think this is an issue with Werkzeug, or something Werkzeug can address.

The Werkzeug development server and debugger is not intended for production use. It is not guaranteed to be particularly secure, stable, or efficient (this is true of Django's development server too).

That being said, you point to `app_iter.close()` not being called, but it is called even if there is an exception. The only way it would not be called is if the worker thread was forcibly closed, which is not something we control.

This seems like a problem with your PostgreSQL setup, or with Django's handling of cleaning up db connections,  as well as your use case of "refreshing the debugger extremely fast". Werkzeug is just wrapping the built-in Python HTTP server, so if it's having trouble keeping up with your requests, then that's an issue with that, not with Werkzeug.

**Comment by Rastopapola** [comment]
Thanks for your fast response.

>This seems like a problem with your PostgreSQL setup 

At first, this was my fear as well, so we checked on everything but found no issue.

>However, I don't think this is an issue with Werkzeug, or something Werkzeug can address.

When I detected, that - in my opinion - something would not run as expected near to `app_iter.close()`, I assumed that the behaviour is Werkzeug related. To make sure, I restarted my dev server using the regular `runserver` instead of the Werkzeug wrapper `runserver_plus` command. 

Surprisingly, all connections are closed properly and the behaviour is not reproduceable anymore, which led me to the conclusion, that something must be going on in the wrapper. 

>The Werkzeug development server and debugger is not intended for production use.

That should - hopefully - be clear for everyone. Stumbling on this behaviour (accidentally) makes you question your code and trying to hunt down a ghost in there.
